content/blogposts/2023/slow-to-simd.md

```diff
@@ -0,0 +1,411 @@
+
+A common method of squeezing more out of our CPUs in situations like this is known as [_loop unrolling_](https://en.wikipedia.org/wiki/Loop_unrolling). The basic idea is to rewrite our loop so more of our relatively-high-latency multiply instructions can execute simultaneously. Additionally, it amortizes the fixed loop costs (increment and compare) across multiple operations.
+
+```go
+func DotUnroll4(a, b []float32) float32 {
+	sum := float32(0)
+	for i := 0; i < len(a); i += 4 {
+		s0 := a[i] * b[i]
+		s1 := a[i+1] * b[i+1]
+		s2 := a[i+2] * b[i+2]
+		s3 := a[i+3] * b[i+3]
+		sum += s0 + s1 + s2 + s3
+	}
+	return sum
+}
+```
+
+In our unrolled code, the dependencies between multiply instructions are removed, enabling the CPU to take more advantage of pipelining. This increases our throughput by 37% compared to our naive implementation.
+
+<Chart
+    rows={[
+        {name: "DotNaive", value: 941900},
+        {name: "DotUnroll4", value: 1286000},
+    ]}
+/>
+
+Note that we can actually improve this slightly more by twiddling with the number of iterations we unroll. On the benchmark machine, 8 seemed to be optimal, but on my laptop, 4 performs best. However, the improvement is quite platform dependent and fairly minimal, so for the rest of the post, I'll stick with an unroll depth of 4 for readability.
+
+## Bounds-checking elimination
+
+In order to keep out-of-bounds slice accesses from being [a security vulnerability](https://en.wikipedia.org/wiki/Heartbleed), the go compiler inserts checks before each read. You can check it out in the [generated assembly](https://go.godbolt.org/z/qT3M7nPGf) (look for `runtime.panic`).
+
+The compiled code makes it look like we wrote somthing like this:
+
+```go
+func DotUnroll4(a, b []float32) float32 {
+	sum := float32(0)
+	for i := 0; i < len(a); i += 4 {
+        if i >= cap(b) {
+            panic("out of bounds")
+        }
+		s0 := a[i] * b[i]
+        if i+1 >= cap(a) || i+1 >= cap(b) {
+            panic("out of bounds")
+        }
+		s1 := a[i+1] * b[i+1]
+        if i+2 >= cap(a) || i+2 >= cap(b) {
+            panic("out of bounds")
+        }
+		s2 := a[i+2] * b[i+2]
+        if i+3 >= cap(a) || >= cap(b) {
+            panic("out of bounds")
+        }
+		s3 := a[i+3] * b[i+3]
+		sum += s0 + s1 + s2 + s3
+	}
+	return sum
+}
+```
+
+In a hot loop like this, even with modern branch prediction, the additional branches per iteration can add up to a pretty significant performance penalty. This is especially true in our case because the inserted jumps limit how much we can take advantage of pipelining.
+
+If we can convince the compiler that these reads can never be out of bounds, it won't insert these runtime checks. This technique is known as "bounds-checking elimination", and the same pattern can be applied to many different memory-safe compiled languages.
```

Comment:
To be pedantic (sorry),  bounds checking also occurs in interpreted languages like Python.

Reply:
Went ahead and removed the confusing reference to compiled languages, and just rephrased to link to BCE recipes in another language.